File: Java/0003_Longest_Substring_Without_Repeating_Characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        if(s.length() == 0) {
            return 0;
        }

        int startIndex = 0;                                             // first index of substring
        int endIndex = 0;                                               // last index of substring
        int maxLength = 0;                                              // max length of substring

        Set<Character> substring = new HashSet<>();                     // HashSet can't contain duplicate values

        while(endIndex < s.length()) {                                  // while end index is less than the input s.length
            if(!substring.contains(s.charAt(endIndex))) {               // if substring doesn't contain duplicate value
                substring.add(s.charAt(endIndex));                      // add that character to substring
                endIndex++;
                maxLength = Math.max(maxLength, substring.size());      // find the max
            } else {                                                    // if substring contains duplicate value
                substring.remove(s.charAt(startIndex));                 // remove start index character
                startIndex++;                                           // so that we can search another substring
            }
        }
        return maxLength;
    }
}

File: JavaScript/0003_Longest_Substring_Without_Repeating_Characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    const seen = new Map();

    var start = 0;
    var max = 0;
    
    for(let i=0; i<s.length; i++) {
        if(seen.has(s[i])) {
            start = Math.max(seen.get(s[i])+1, start);
        }

        seen.set(s[i], i);
        max = Math.max(i-start+1, max);
    }
    return max;
};
